### ISSUE TITLE
[java] Rule to detect stateful enums

### ISSUE THREAD

**Issue by oowekyala** [pmd/pmd#2916]
**Proposed Rule Name:** StatefulEnum

**Proposed Category:** Design

**Description:** Flags enums that look like they have mutable state, like 
* setters
  * Possibly, any `void` method indicates that there is some statefulness (wdyt?)
* non-final instance fields
* synchronization constructs

**Code Sample:** This should include code, that should be flagged by the rule. If possible, the "correct" code
according to this new rule should also be demonstrated.

```java
enum Foo {
   A;
  private String name;

  public void setName(String name) {
     this.name = name;
  }
}
```

```java
enum Foo {
   A;
  private String name;

  public synchronized void getSomething() {
     // synchronized? If the enum constant is really immutable, then no data race can occur
     // So the synchronized hints that there is something fishy... or at least it is redundant
  }
}
```

**Possible Properties:**

* Maybe optionally checking static fields/methods?
  * Eg, if there is a static concurrent map
  * Eg, if there are static setters/void methods
  * But this looks like it could be generalized to a "MutableStaticState" rule, which does not only check enums
*   What else ??

**Comment by Fernal73** [comment]
There also happens to be a similar Sonar rule:-

https://rules.sonarsource.com/java/RSPEC-3066

**Comment by Fernal73** [comment]
You could have setters in implemented interfaces but you'd probably define the method as abstract and override it for each instance.
